### PR TITLE
fix: permit appropriate computed member expressions and prototype access

### LINKED ISSUE #533: Access to `Promise.prototype.%somethingStandard%` trigger `promise/spec-only`
```
x:y  error  Avoid using non-standard 'Promise.prototype'  promise/spec-only
```

### LINKED ISSUE #534: Access to `Promise` methods in a way different from dot trigger `spec-only`
```js
Promise['all'];  // error  Avoid using non-standard 'Promise.undefined'  promise/spec-only
var method = 'all';
Promise[method]; // error  Avoid using non-standard 'Promise.method'     promise/spec-only
```

### PR DESCRIPTION
**Pull request by brettz9**
**What is the purpose of this pull request?**

- [x] Bug fix

**What changes did you make? (Give an overview)**

Permits appropriate computed member expressions such as `Promise['all']` and suitable prototype access (e.g., `Promise.prototype.catch`) but errs on detectable problems (e.g., `Promise.prototype.done`).

Fixes #533, Fixes #534.

### CODE REVIEW

File: rules/spec-only.js [hunk 1]
```diff
@@ -35,14 +37,19 @@ module.exports = {
...
         if (
           node.object.type === 'Identifier' &&
+          (!node.computed || node.property.type === 'Literal') &&
           node.object.name === 'Promise' &&
-          !(node.property.name in PROMISE_STATICS) &&
+          ((node.property.name && !PROMISE_STATICS.has(node.property.name)) ||
+            (node.property.value &&
+              !PROMISE_STATICS.has(node.property.value))) &&
+          (node.property.name !== 'prototype' ||
+            !PROMISE_INSTANCE_PROPS.has(node?.parent?.property?.name)) &&
```
Comment (zloirock) [inline rules/spec-only.js:46]: In the future, to `Promise.prototype` could be added new methods, like it was with `.finally`. And in this case here will be a situation like with static `Promise.try` now - it's still not in the spec, but it's on stage 3 and shipped in the latest Chrome, for usage it should be added to `allowedMethods`. But here is no similar check for instance methods.

Comment (brettz9) [reply rules/spec-only.js]: Thanks, I've added an instance method check.

File: rules/spec-only.js [hunk 2]
```diff
@@ -35,14 +37,20 @@ module.exports = {
...
           node.object.name === 'Promise' &&
-          !(node.property.name in PROMISE_STATICS) &&
-          !allowedMethods.includes(node.property.name)
+          ((node.property.name && !PROMISE_STATICS.has(node.property.name)) ||
+            (node.property.value &&
+              !PROMISE_STATICS.has(node.property.value))) &&
+          (node.property.name !== 'prototype' ||
+            (!PROMISE_INSTANCE_METHODS.has(node?.parent?.property?.name) &&
+              !allowedMethods.includes(node?.parent?.property?.name))) &&
+          !allowedMethods.includes(node.property.name ?? node.property.value)
```
[the remark this answers is not in the document]
Comment (brettz9) [reply rules/spec-only.js]: A very helpful function and examples, thanks! Added... I have allowed bare `prototype` access, however, as I think we won't know how it is being used (like with `Promise[method]`).

File: __tests__/spec-only.js [hunk 3]
```diff
@@ -9,7 +9,10 @@ ruleTester.run('spec-only', rule, {
     'Promise.resolve()',
     'Promise.reject()',
     'Promise.all()',
+    'Promise["all"]',
+    'Promise[method];',
```
[the remark this answers is not in the document]
Comment (brettz9) [reply __tests__/spec-only.js:13]: Thanks for the examples. Incorporated. I do think it should be permitted as it could I think be used benignly.

**Review by scagood — APPROVED** [review]
LGTM :muscle:

**Review by ota-meshi — APPROVED** [review]

**Pull request merged**